src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java

```diff
@@ -252,84 +259,115 @@ private MqttConnectOptions getConnectionOptions() throws KeyStoreException {
+        if (mqttClientInternal.isConnected()) {
             LOGGER.atInfo()
                     .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
                     .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
-                    .log("Connected to broker");
+                    .log("Already connected, skipping connect");
+            return;
         }
-    }
-
-    private void reconnectAndResubscribe() {
-        int waitBeforeRetry = MIN_WAIT_RETRY_IN_SECONDS;
 
-        while (!mqttClientInternal.isConnected() && !Thread.currentThread().isInterrupted()) {
-            try {
-                // TODO: Clean up this loop
-                doConnect();
-            } catch (MqttException | KeyStoreException e) {
-                if (Utils.getUltimateCause(e) instanceof InterruptedException) {
-                    // paho doesn't reset the interrupt flag
-                    LOGGER.atDebug().log("Interrupted during reconnect");
-                    Thread.currentThread().interrupt();
-                    return;
-                }
-
-                LOGGER.atDebug().setCause(e)
-                        .log("Unable to connect. Will be retried after {} seconds", waitBeforeRetry);
+        try {
+            RetryUtils.runWithRetry(mqttExceptionRetryConfig, () -> {
+                LOGGER.atInfo()
+                        .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
+                        .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
+                        .log("Connecting to broker");
+                MqttConnectOptions opts = getConnectionOptions();
                 try {
-                    Thread.sleep(waitBeforeRetry * 1000);
-                } catch (InterruptedException er) {
-                    Thread.currentThread().interrupt();
-                    LOGGER.atDebug().log("Interrupted during reconnect");
-                    return;
+                    mqttClientInternal.connect(opts);
+                } catch (MqttException e) {
+                    if (Utils.getUltimateCause(e) instanceof InterruptedException) {
+                        // paho doesn't reset the interrupt flag
+                        LOGGER.atDebug().log("Interrupted during reconnect");
+                        Thread.currentThread().interrupt();
+                        throw (InterruptedException) Utils.getUltimateCause(e);
+                    }
+                    throw e;
                 }
-                waitBeforeRetry = Math.min(2 * waitBeforeRetry, MAX_WAIT_RETRY_IN_SECONDS);
-            }
+                return null;
+            }, "connect-mqtt-topic", LOGGER);
+            LOGGER.atInfo()
+                    .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
+                    .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
+                    .log("Connected to broker");
+        } catch (InterruptedException e) {
+            throw e;
+        } catch (Exception e) {
+            // TODO considering erroring the component, or other recovery options
```

Comment:
retrying?

Reply:
yeah that's fair, i suppose doesn't hurt to retry everything